Fixes #16313 and Adds back M1.3 And M1.4 (Blog Integration)

Rijuta-s opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->
 Changes in this PR - [846f4f9](https://github.com/oppia/oppia/pull/16323/commits/846f4f908514d0cf8045aa0028249e4e29319e5f)
1. This PR fixes #16313 
2. This PR does the following: This PR fixes the FE flake and adds back reverted PR #https://github.com/oppia/oppia/pull/16040

Link to FE tests: https://github.com/Rijuta-s/frontend-tests-flake-check-oppia/actions/workflows/frontend_tests_flake_check.yml
## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
FE tests pass on PR  for 50 times:
![image](https://user-images.githubusercontent.com/68547101/196011716-93420eff-93ba-4ce5-8a73-a9baa06770bb.png)

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

seanlip committed:
Try to fix blog page flake.

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into develop

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into develop

Rijuta-s committed:
Building blog post search service

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog_post_search

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog_post_search

Rijuta-s committed:
Fixing linting issue

Rijuta-s committed:
Self review

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog_post_search

Rijuta-s committed:
fixing search service and adding tests

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog_post_search

Rijuta-s committed:
fixing backend test shard

Rijuta-s committed:
fixing backend test

Rijuta-s committed:
fixing backend test

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog_post_search

Rijuta-s committed:
fixing backend test

Rijuta-s committed:
Applying SUggestions

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog_post_search

Rijuta-s committed:
Fixing lint errors

Rijuta-s committed:
Fixing lint errors

Rijuta-s committed:
Applying Suggestions

Rijuta-s committed:
Applying Suggestions

Rijuta-s committed:
Applying Suggestions

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog_post_search

Rijuta-s committed:
Merge remote-tracking branch upstream/develop into blog_post_search

Rijuta-s committed:
Fixing Lint issues

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog_post_search

Rijuta-s committed:
Merge remote-tracking branch 'origin/blog_post_search' into blog-home-page

Rijuta-s committed:
blog homepage and postpage

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog-home-page

Rijuta-s committed:
fixing lint issues

Rijuta-s committed:
fixing lint issues

Rijuta-s committed:
fixing lint issues

Rijuta-s committed:
fixing frontend checks

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog-home-page

Rijuta-s committed:
fixing frontend checks

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog-home-page

Rijuta-s committed:
self review

Rijuta-s committed:
self review

Rijuta-s committed:
fixing backend test

Rijuta-s committed:
fixing frontend test

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog-home-page

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog-home-page

Rijuta-s committed:
fixing backend coverage and descriptions

Rijuta-s committed:
fixing backend coverage and descriptions

Rijuta-s committed:
fixing frontend test

Rijuta-s committed:
applying suggestions

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog-home-page

Rijuta-s committed:
minor fixes and Applying Suggestions

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog-home-page

Rijuta-s committed:
minor fixes and Applying Suggestions

Rijuta-s committed:
minor fix

Rijuta-s committed:
Merge branch 'develop' into blog-home-page

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog-home-page

Rijuta-s committed:
fixing FE flake

gp201 reviewed (approved):
LGTM

gp201 commented:
@seanlip can we go ahead and merge this since the rest of the changes have already been approved?

seanlip commented on core/templates/pages/blog-home-page/tag-filter/tag-filter.component.spec.ts:
```diff
@@ -63,40 +62,36 @@ describe('Tag Filter component', () => {
+        }
+      } as ElementRef;
+      component.selectTag(({ option: { viewValue: 'tag3'}}));
+      tick(1600);
```
This needs justification. Why 1600 and how was this value determined?

Ditto below.

Rijuta-s replied:
Changed to 1501ms after discussion with @gp201 . A debounce time of 1500 ms is applied in tag filter, i.e only if no changes in the tag filter are done by the user for 1500ms, search with the applied filter will be executed. 1 millisecond extra to be on safer side and avoid flaking!

seanlip replied:
To be clear, I don't care too much about the exact value. My main concern is that a fellow developer who's reading this code will have no idea where these numbers come from. I think you should add code comments to explain. Ideally we have a common constant which both pieces of code refer to, so that if 1500 is changed then the test gets the update automatically too.

Note also that 1501 ms is 1 millisecond extra, and not 1 second extra, so I'm not sure whether you made the change that you wanted to make.

/cc @gp201

gp201 replied:
@seanlip For frontend tests, we usually don't have common constants for the setTimeout/debounce time that is used in both `*.spec.ts` and `*.ts` files. But we could enforce this method throughout the codebase.

We also add comments only if there are multiple `tick` calls or if the `tick` is accounting for a call not present in the file directly being tested. Since the tick is usually called right after the function call, we expect developers to be able to easily find what the `tick` call is accounting for.

In some frontend tests, we observe that the function call is not completed when the time set in the `tick` is equal to the setTimeout/debounce time, and we would have to add 1 millisecond to solve it.

seanlip replied:
OK, yes, let's use a common constant. I imagine we can use the same debounce time everywhere.

I think any presence of "magic numbers" in the codebase needs at least an explanatory comment, though using a constant is better and that shouldn't need a comment.

gp201 replied:
@seanlip we can't use the same denounce time for the entire codebase. Since the time is selected carefully depending on the situation.

Sorry, I wasn't clear enough. There are 2 reasons we don't create constants like that:
- if we were to declare a const in the component file we won't be able to access it in the spec file. The only way to access it would be to define it as `this.var = ...` which is not ideal. (Note that this is possible but it is not going to be a good approach.)
- as stated above if we were to define the same value in the global constants file then we would a large number of values since each value is selected based on the situation.

seanlip replied:
Can you explain why "the time is selected carefully depending on the situation"? What factors go into selecting the correct debounce time and what would break if it was, say, 1.5 seconds everywhere?

For the second point, how many constants are there? I think having lots of individual independent constants for things like this is a maintenance burden so I would like to dig more into why further unification is not possible.

Rijuta-s replied:
@seanlip  @gp201 Added a comment and a constant (specific to the blog-home-page as ther are differect values of debounce timeout being used in different places). PTAL!

(Also sorry for the typo above, I meant 1 ms --not 1 second)

seanlip reviewed (changes requested):
Left one comment.

@gp201 re:

> @seanlip can we go ahead and merge this since the rest of the changes have already been approved?

Yes, on one condition: could you or Rijuta please explain what steps are we taking to prevent this sort of thing from happening again? (I know you've mentioned some of this in a groupchat, but we should document these so that anyone in the dev community can refer to them, in case this recurs.) We should also file an issue to track the fix if one is needed.

heyimShivam committed:
done

heyimShivam committed:
done

gp201 commented:
The following steps are going to be taken to prevent flakes from entering again:

1. Bulk run tests after Pr is merged. (This feature had already been implemented via https://github.com/oppia/oppia/pull/16174)
2. muti-run spec files that have been modified prior to merge (An issue detailing the plan for this will be opened soon).
3. As soon as the fronted tests fail on develop the corresponding PR has to reverted (ASAP: within a few hours of being detected).

gp201 commented:
@seanlip PTAL

Rijuta-s committed:
Update tag-filter.component.spec.ts

heyimShivam committed:
done

Rijuta-s committed:
Merge branch 'develop' into blog-home-page

heyimShivam committed:
updated

heyimShivam committed:
done

heyimShivam committed:
done

heyimShivam committed:
coverage

heyimShivam committed:
updated

heyimShivam committed:
done

heyimShivam committed:
updated

heyimShivam committed:
updated

heyimShivam committed:
up down

heyimShivam committed:
done4

Rijuta-s commented:
@gp201 Bulk frontend tests seem to fail due to flake in 'context service'. PTAL!

gp201 commented:
@Rijuta-s @heyimShivam has opened a PR to fix that flake: https://github.com/oppia/oppia/pull/16337

heyimShivam committed:
done

heyimShivam committed:
f

heyimShivam committed:
done

heyimShivam committed:
done

heyimShivam committed:
anoher

heyimShivam committed:
nice

heyimShivam committed:
nice one

heyimShivam committed:
done

Rijuta-s committed:
Merge branch 'frontend' of https://github.com/heyimShivam/oppia into blog-home-page

Rijuta-s committed:
Merge branch 'develop' into blog-home-page

seanlip reviewed (changes requested):
@Rijuta-s @gp201 Please see https://github.com/oppia/oppia/pull/16323#discussion_r999840954. Thanks!

gp201 commented:
@seanlip [replied](https://github.com/oppia/oppia/pull/16323#discussion_r1000199275) to the comment. Thanks!

seanlip commented:
Thanks -- I've replied.

gp201 commented:
@seanlip PTAL. I've replied

Rijuta-s committed:
Merge remote-tracking branch 'upstream/develop' into blog-home-page

Rijuta-s committed:
applying suggestions

seanlip reviewed (approved):
OK, LGTM -- thanks! Once CI passes lmk and I can merge this PR.

Rijuta-s commented:
@seanlip All checks are passing now! Please merge the PR!

seanlip commented:
Hi @Rijuta-s, I think I am confused. What's the connection between this and #16299? I might have mistaken one PR for the other since both say "add back M1.3 and M1.4".

Rijuta-s commented:
> Hi @Rijuta-s, I think I am confused. What's the connection between this and #16299? I might have mistaken one PR for the other since both say "add back M1.3 and M1.4".

Hi @seanlip , #16299 adds back PR #16123 (it addressed issues from product review done by Diana) which was also reverted back due to introduction of bug into RTE. This PR(#16123) was stacked over M1.3 and M1.4. Therefore, I just pulled changes of this PR in #16299 to remove merge - conflicts from #16299. #16323 adds back M1.3 and M1.4.

seanlip commented:
Ah ok, thanks. For stacked PRs in the future, here is what to do: keep the title of the stacked PR restricted to *only* the stacked PR's changes. In the description, explain that this PR is stacked on PR XXX and reviewers should start reviewing from commit XYZ. This will help address the confusion.
